_utils/firefox_download.py

```diff
@@ -0,0 +1,133 @@
+            os.makedirs(self.out_dir)
+
+        # Create an etags cache INI file, if one doesn't already exist.
+        if not os.path.isfile(self.cache_path):
+            with open(self.cache_path, 'w') as file:
+                cache = IniHandler()
+                cache.config.add_section('etags')
+                cache.config.add_section('cached')
+                cache.config.write(file)
+                file.close()
+
+    def download_all(self):
+        """
+        Loop over each channel in the config INI file and download the latest
+        build.
+        """
+        IniHandler.banner('DOWNLOADING FIREFOXES')
+
+        for channel in self.config.sections():
+            self.download_channel(channel)
+
+    def download_channel(self, channel):
+        """
+        Download a specific channel of Firefox (ie: nightly, aurora, beta, ...)
+        """
+        # Get the full download URL and output filename from the config.
+        download_filename = self.config.get(channel, 'DOWNLOAD_FILENAME')
+        download_url = self.config.get(channel, 'DOWNLOAD_URL')
+        out_file = os.path.join(self.out_dir, download_filename)
+
+        # Attempt to download the latest version of the Firefox binary for
+        # the current channel.
+        self.download_file_from_url(download_url, out_file, channel)
+
+    def download_file_from_url(self, download_url, out_file, channel):
+        """
+        Download the specified channel Firefox build, using an "intelligent"
+        cache, using the archive's current `etag` as a unique identifier.
+        """
+        with open(out_file, 'wb') as file:
+            try:
+                contents = urllib.request.urlopen(download_url)
+                etag = contents.headers['etag']
+                was_cached = False
+
+                # We found this channel in the cache, but not sure if it is
+                # recent.
+                if self.cache.config.has_option('etags', channel):
+                    # Etags matched, we probably have the latest version.
+                    if etag == self.cache.get('etags', channel):
+                        was_cached = True
+
+                # Update the etag cache with the latest etag fingerprint.
+                self.cache.set('etags', channel, etag)
+                self.cache.set('cached', channel, was_cached)
+
+                if not was_cached:
+                    print(('Downloading [{0}] from {1}'.format(channel, download_url)))
+                    file.write(contents.read())
+                    file.close()
+
+                else:
+                    print(('[{0}] etag was cached, skipping download.'.format(channel)))
+
+            except urllib.error.HTTPError as err:
+                # For some reason the download failed. Possibly due to a config
+                # mismatch in the INI files. :shrug:
+                # TODO: Currently we fail silently here. Fail harder?
+                print(('\t', err))
+
```

Comment:
Ideally we'd do something like a retry, but occasionally we've seen the download server offline for up to an hour or so so may not be as useful.  Also, unlike for client-testing, it's seldom urgent for us to have the latest nightly every single day (though that may not always be true for services like shavar which are tightly integrated with client).

for now lets exit with something like 
exit('ERROR: {0}'. format(err))
if we discover the download server is very flaky, then we may consider in future allowing job to continue but flagging the script to notify jenkins to fail the job with an error code